OPHD/States/MapViewStateTurn.cpp

```diff
@@ -334,6 +341,42 @@ void MapViewState::updateResidentialCapacity()
 }
 
 
+void MapViewState::updateBiowasteRecycling()
+{
+	auto& residences = NAS2D::Utility<StructureManager>::get().structureList(Structure::StructureClass::Residence);
+	auto& recyclingFacilities = NAS2D::Utility<StructureManager>::get().structureList(Structure::StructureClass::Recycling);
+
+	if (residences.empty() || recyclingFacilities.empty()) { return; }
+
+	auto residenceIterator = residences.begin();
+	for (auto recyclingFacility : recyclingFacilities)
+	{
+		int count = 0;
+		
+		if (!recyclingFacility->operational()) { continue; } // Consider a different control structure
+
+		Recycling* recycling = static_cast<Recycling*>(recyclingFacility);
+
+		for (; residenceIterator != residences.end(); ++residenceIterator)
+		{
+			Residence* residence = static_cast<Residence*>(*residenceIterator);
+			residence->pullWaste(recycling->wasteProcessingCapacity());
+			
+			++count;
+			
+			if (count >= recycling->residentialSupportCount())
+			{
+				break;
+			}
+		}
+
+		// We got to the end of the residence list, don't waste time iterating over
+		// any remaining recycling facilities.
+		if (residenceIterator == residences.end()) { break; }
+	}
```

Comment:
Maybe the loop should be on the `count` variable, with an early exit via `return` if the end of the residences list is reached. That would avoid the need for an `if` test after the `for` loop.

```cpp
for (auto recyclingFacility : recyclingFacilities) {
  for (int count = 0; count < recycling->residentialSupportCount(); ++count) {
    if (residenceIterator == residences.end()) {
      return; // No more residences, so don't waste time iterating over remaining recycling facilities
    }
    Residence* residence = static_cast<Residence*>(*residenceIterator);
    residence->pullWaste(recycling->wasteProcessingCapacity());
  }
}
```

Reply:
I've noticed those `operational()` checks in a few places. Maybe it would be helpful to have a method that returns a list of operational structures of a given type.

Perhaps that can tie in to the earlier idea from maybe a few months back about having a template method to return a list of structures of a given type, so the casting isn't needed.

Reply:
Blah, managed to miss this comment before merging.

I like the suggestion. Reminds me to get a second set of eyes on code.

Candidate for refactoring later on :D